File: frontend/src/shared/types/workspace.ts
export type NodeType = "folder" | "file";

export type TabKind = "file" | "folder";

export interface WorkbenchTab {
  kind: TabKind;
  path: string;
}

export function tabKey(tab: WorkbenchTab): string {
  return `${tab.kind}:${tab.path}`;
}

export function tabsEqual(a: WorkbenchTab | null, b: WorkbenchTab | null): boolean {
  if (!a || !b) {
    return a === b;
  }
  return a.kind === b.kind && a.path === b.path;
}

export interface TreeNode {
  id: string;
  name: string;
  path: string;
  type: NodeType;
  hasChildren: boolean;
  children?: TreeNode[] | null;
}

export interface TreeResponse {
  parentPath: string;
  nodes: TreeNode[];
}

export interface WorkspaceInfo {
  root: string;
  name: string;
  docsRoot: string;
}

export interface CreateFolderRequest {
  parentPath: string;
  name: string;
}

export interface CreateDocumentRequest {
  parentPath: string;
  name: string;
  templateType?: string;
}

export interface UpdateNodeRequest {
  path: string;
  newName?: string;
  newParentPath?: string;
}

export interface DeleteNodeResponse {
  deleted: string;
  type: NodeType;
}

export interface ApiErrorBody {
  code: string;
  message: string;
  details?: Record<string, unknown>;
}

File: backend/api/src/kew_api/api/v1/workspace.py
"""Workspace and documentation tree endpoints."""

from __future__ import annotations

from fastapi import APIRouter, Query

from kew_api.api.deps import WorkspaceDep
from kew_api.schemas.workspace import (
    CreateDocumentRequest,
    CreateFolderRequest,
    DeleteNodeResponse,
    TreeNode,
    TreeResponse,
    UpdateNodeRequest,
    WorkspaceInfo,
)

router = APIRouter(prefix="/workspace", tags=["workspace"])


@router.get("", response_model=WorkspaceInfo)
def get_workspace(service: WorkspaceDep) -> WorkspaceInfo:
    return service.get_workspace_info()


@router.get("/tree", response_model=TreeResponse)
def get_tree(
    service: WorkspaceDep,
    parent_path: str = Query(default="", description="Relative path under docs/"),
    depth: int = Query(default=1, ge=1, le=5),
) -> TreeResponse:
    return service.list_tree(parent_path=parent_path, depth=depth)


@router.post("/folders", response_model=TreeNode, status_code=201)
def create_folder(body: CreateFolderRequest, service: WorkspaceDep) -> TreeNode:
    return service.create_folder(parent_path=body.parent_path, name=body.name)


@router.post("/documents", response_model=TreeNode, status_code=201)
def create_document(body: CreateDocumentRequest, service: WorkspaceDep) -> TreeNode:
    return service.create_document(
        parent_path=body.parent_path,
        name=body.name,
        template_type=body.template_type,
    )


@router.patch("/nodes", response_model=TreeNode)
def update_node(body: UpdateNodeRequest, service: WorkspaceDep) -> TreeNode:
    return service.update_node(
        body.path,
        new_name=body.new_name,
        new_parent_path=body.new_parent_path,
    )


@router.delete("/nodes", response_model=DeleteNodeResponse)
def delete_node(
    service: WorkspaceDep,
    path: str = Query(min_length=1, description="Relative path under docs/"),
) -> DeleteNodeResponse:
    return service.delete_node(path)
